src/lib/OpenEXR/ImfContext.cpp

```diff
@@ -0,0 +1,866 @@
+
+#include "openexr.h"
+
+#include "Iex.h"
+
+// TODO: remove these once we've cleared the legacy stream need
+#include "ImfIO.h"
+#include "ImfStdIO.h"
+#include <mutex>
+
+#include <string.h>
+
+#include <ImfBoxAttribute.h>
+#include <ImfChannelListAttribute.h>
+#include <ImfChromaticitiesAttribute.h>
+#include <ImfCompressionAttribute.h>
+#include <ImfCompressor.h>
+#include <ImfDeepImageStateAttribute.h>
+#include <ImfDoubleAttribute.h>
+#include <ImfDwaCompressor.h>
+#include <ImfEnvmapAttribute.h>
+#include <ImfFloatAttribute.h>
+#include <ImfFloatVectorAttribute.h>
+#include <ImfIDManifestAttribute.h>
+#include <ImfIntAttribute.h>
+#include <ImfKeyCodeAttribute.h>
+#include <ImfLineOrderAttribute.h>
+#include <ImfMatrixAttribute.h>
+#include <ImfOpaqueAttribute.h>
+#include <ImfPreviewImageAttribute.h>
+#include <ImfRationalAttribute.h>
+#include <ImfStringAttribute.h>
+#include <ImfStringVectorAttribute.h>
+#include <ImfTileDescriptionAttribute.h>
+#include <ImfTimeCodeAttribute.h>
+#include <ImfVecAttribute.h>
+
+OPENEXR_IMF_INTERNAL_NAMESPACE_SOURCE_ENTER
+
+namespace
+{
+
+class MemAttrStream : public OPENEXR_IMF_NAMESPACE::IStream
+{
+public:
+    MemAttrStream (const exr_attr_opaquedata_t *opaque)
+        : IStream ("<mem_attr>"),
+          _data (static_cast<char*> (opaque->packed_data)),
+          _sz (static_cast<uint64_t> (opaque->size)),
+          _pos (0)
+    {}
+
+    ~MemAttrStream () override {}
+
+    bool isMemoryMapped () const override { return true; }
+
+    bool     read (char c[/*n*/], int n) override
+    {
+        if (_pos >= _sz && n != 0)
+            throw IEX_NAMESPACE::InputExc ("Unexpected end of file.");
+
+        uint64_t n2     = n;
+        bool     retVal = true;
+
+        if (_sz - _pos <= n2)
+        {
+            n2     = _sz - _pos;
+            retVal = false;
+        }
+
+        memcpy (c, _data + _pos, n2);
+        _pos += n2;
+        return retVal;
+    }
+
+    char*    readMemoryMapped (int n) override
+    {
+        if (_pos >= _sz)
+            throw IEX_NAMESPACE::InputExc ("Unexpected end of file.");
+
+        if (_pos + n > _sz)
+            throw IEX_NAMESPACE::InputExc ("Reading past end of file.");
+
+        char* retVal = _data + _pos;
+        _pos += n;
+        return retVal;
+    }
+    uint64_t tellg () override { return _pos; }
+    void     seekg (uint64_t pos) override { _pos = pos; }
+    void     clear () override {}
+
+private:
+    char*    _data;
+    uint64_t _sz;
+    uint64_t _pos;
+};
+
+} // namespace
+
+////////////////////////////////////////
+
+Context::Context ()
+    : _ctxt (new exr_context_t, [] (exr_context_t* todel) {
+        exr_finish (todel);
+        delete todel;
+    })
+{
+    *_ctxt = nullptr;
+}
+
+////////////////////////////////////////
+
+Context::Context (const char* filename, const ContextInitializer& ctxtinit, read_mode_t)
+    : Context()
+{
+    exr_result_t rv;
+    if (*_ctxt)
+    {
+        THROW (
```

Comment:
how can we end up in this THROW?

Reply:
it's a fair question - was legacy from how I initially defined this stuff... less code, less bugs are incoming!